File: exercise/python/day24/usecases/toy_delivery_use_case.py
from typing import Union

from domain.core.error import Error
from domain.toy import Toy
from domain.toy_repository import ToyRepository
from usecases.deliver_toy import DeliverToy


class ToyDeliveryUseCase:
    def __init__(self, repository: ToyRepository):
        self.repository = repository

    def handle(self, deliver_toy: DeliverToy) -> Union[Error, None]:
        toy = self.repository.find_by_name(deliver_toy.desired_toy)
        if toy is None:
            return Error.an_error(f"Oops we have a problem... we have not built the toy: {deliver_toy.desired_toy}")
        result = self.reduce_stock(toy)
        if isinstance(result, Error):
            return result
        return None

    def reduce_stock(self, toy: Toy) -> Union[Error, Toy]:
        result = toy.reduce_stock()
        if isinstance(result, Toy):
            self.repository.save(result)
        else:
            self.repository.save(toy)
        return result

File: exercise/java/day24/src/main/java/usecases/ToyDeliveryUseCase.java
package usecases;

import domain.Toy;
import domain.ToyRepository;
import domain.core.Error;
import domain.core.Unit;
import io.vavr.control.Either;

import static domain.core.Error.anError;

public class ToyDeliveryUseCase {
    private final ToyRepository repository;

    public ToyDeliveryUseCase(ToyRepository repository) {
        this.repository = repository;
    }

    public Either<Error, Unit> handle(DeliverToy deliverToy) {
        return repository
                .findByName(deliverToy.desiredToy())
                .toEither(() -> errorFor(deliverToy))
                .flatMap(this::reduceStock)
                .map(ignored -> Unit.DEFAULT);
    }

    private Either<Error, Toy> reduceStock(Toy toy) {
        return toy.reduceStock()
                .peek(repository::save)
                .peekLeft(left -> repository.save(toy))
                .map(updatedToy -> toy);
    }

    private static Error errorFor(DeliverToy deliverToy) {
        return anError("Oops we have a problem... we have not built the toy: " + deliverToy.desiredToy());
    }
}